docs/HowToCertificatesRotation.md

```diff
@@ -0,0 +1,473 @@
+      "expiresTimestamp": "2121-07-05T09:19:43.000+0000",
+      "issuerName": "CN=ca.nuodb.com, OU=Eng, O=NuoDB, L=Boston, ST=MA, C=US, SERIALNUMBER=67890",
+      "subjectName": "CN=*.nuodb.svc.cluster.local, OU=Eng, O=NuoDB, L=Boston, ST=MA, C=US, SERIALNUMBER=67890"
+    }
+  },
+  "serverTrusted": {
+    "admin-nuodb-cluster0-0": [
+      "ca_prime",
+      "nuocmd_prime"
+    ],
+    "admin-nuodb-cluster0-1": [
+      "ca_prime",
+      "nuocmd_prime"
+    ]
+  },
+  "trustedCertificates": {
+    "ca_prime": {
+      "caPathLength": 2147483647,
+      "certificatePem": "-----BEGIN CERTIFICATE-----...-----END CERTIFICATE-----",
+      "expires": 4781150383000,
+      "expiresTimestamp": "2121-07-05T09:19:43.000+0000",
+      "issuerName": "CN=ca.nuodb.com, OU=Eng, O=NuoDB, L=Boston, ST=MA, C=US, SERIALNUMBER=67890",
+      "subjectName": "CN=ca.nuodb.com, OU=Eng, O=NuoDB, L=Boston, ST=MA, C=US, SERIALNUMBER=67890"
+    },
+    "nuocmd_prime": {
+      "caPathLength": -1,
+      "certificatePem": "-----BEGIN CERTIFICATE-----...-----END CERTIFICATE-----",
+      "expires": 4781151591000,
+      "expiresTimestamp": "2121-07-05T09:39:51.000+0000",
+      "issuerName": "CN=nuocmd.nuodb.com, OU=Eng, O=NuoDB, L=Boston, ST=MA, C=US, SERIALNUMBER=67890",
+      "subjectName": "CN=nuocmd.nuodb.com, OU=Eng, O=NuoDB, L=Boston, ST=MA, C=US, SERIALNUMBER=67890"
+    }
+  }
+}
+```
+
+## NuoDB TLS Key Rotation
+
+To renew the domain key pair certificates and the server keystore, you can either create new keystore files and certificates on your own or create them using NuoDB commands.
+All examples below will be using a helper pod started with NuoDB image and `nuocmd` command line tools.
+
+```bash
+PASSWD=changeMe
+mkdir /tmp/keys
+
+kubectl run generate-nuodb-certs \
+  --image nuodb/nuodb-ce \
+  --env="PASSWD=changeMe" \
+  --command -- 'tail' '-f' '/dev/null'
+
+kubectl exec -ti generate-nuodb-certs -- \
+  mkdir -p /tmp/keys
+```
+
+### Rotate CA Certificate
+
+Generate new CA certificate:
+
+```bash
+kubectl exec -ti generate-nuodb-certs -- \
+  nuocmd create keypair \
+    --keystore /tmp/keys/ca.p12 --store-password "$PASSWD" \
+    --dname "CN=ca.nuodb.com, OU=Eng, O=NuoDB, L=Boston, ST=MA, C=US, SERIALNUMBER=67890" \
+    --validity 36500 --ca
+
+kubectl exec -ti generate-nuodb-certs -- bash -c \
+  'nuocmd show certificate \
+    --keystore /tmp/keys/ca.p12 --store-password "$PASSWD" \
+    --cert-only > /tmp/keys/ca.cert'
+```
+
+If the CA certificate is owned by a public CA, then follow the official steps to obtain the new certificate.
+
+The newly created keystore files can be copied on the client machine and then on some AP pod.
+
+```bash
+kubectl cp generate-nuodb-certs:/tmp/keys/. /tmp/keys
+
+kubectl cp /tmp/keys/ca.cert nuodb/admin-nuodb-cluster0-0:/tmp/ca.cert
+```
+
+Add the new CA certificate to the truststore of all admin and database processes:
+
+```bash
+kubectl exec -ti admin-nuodb-cluster0-0 -- \
```

Comment:
I think there is an overlap here that needs to be explained between copying _ca.cert_ to the pod and running _nuocmd add trusted-certificate_, and the earlier steps of _kubectl create secret generic nuodb-ca-cert-renewed_.

A long time ago, I remember having to create a combined ca.cert with both old and new in the k8s secret / vault before running _add trusted-certificate_ but that was for the shared model.

Reply:
I'm not sure that I understand the first part. The `nuodb-ca-cert-renewed` is the actual certificate that `nuocmd` uses to verify the admin. In the steps we use the exact same `ca.cert` file for both invoking `nuocmd add trusted-certificate` and creating the secret.

> A long time ago, I remember having to create a combined ca.cert with both old and new in the k8s secret / vault before running add trusted-certificate but that was for the shared model.

I believe that you are referring to updating the certificate that the SQL clients use instead of the one that is added to admin truststore? Step 2 from https://doc.nuodb.com/nuodb/latest/deployment-models/physical-or-vmware-environments-with-nuodb-admin/domain-operations/enabling-tls-encryption/rotating-key-pair-certificates/#_rotating_keys_for_the_shared_admin_key_trust_model. The new CA certificate should be added to the SQL clients truststore, which is documented below. (I will also include an example).

I think that there is no need to combine the old and the new CA certificate for `nuocmd add trusted-certificate` to work. @adriansuarez Can you confirm?

Another question is should the `nuodb-ca-cert-renewed` include the old CA certificate. I think _no_ because `nuodocker start admin` doesn't use the REST. The database chart is upgraded only after the admin certificates are fully rotated so there will be no admin pod running with the old key pair.
I've tested the steps with just 2 admins so I will double-check them with 3 to be sure.

Reply:
> I'm not sure that I understand the first part. The nuodb-ca-cert-renewed is the actual certificate that nuocmd uses to verify the admin. In the steps we use the exact same ca.cert file for both invoking nuocmd add trusted-certificate and creating the secret.

Sorry my statement was not clear.  I agree they are the same thing - hence the overlap I meant is that a reader might ask _Why am I supplying the same file via a secret as well as having to copy into the pod manually (to a different location) and use it to run nuocmd add trusted-certificate_.  We know why but from a readers perspective it seems like duplication.

> I think that there is no need to combine the old and the new CA certificate for nuocmd add trusted-certificate to work.

Yes step 2 there looks familiar.  Before, I combined it only for the secret/vault value, not for the nuocmd command (that only used the new one).  Step 2 combines it for the file in /etc/nuodb/keys/ (which is the file in the secret / vault) before running _nuocmd add trusted-certificate_ for the new certificate only.  I only raise it as a checkpoint, it has been a while now since I played with vault rotation.

Reply:
@acabrele which file are you referring to, `ca.cert`, `nuoadmin.p12`, or `nuoadmin-truststore.p12`? Of those, only the K8s Secrets for `ca.cert` and `nuoadmin.p12` should be updated. `nuoadmin-truststore.p12` should not be updated manually if we are relying on `nuocmd add trusted-certificates` to propagate the new trusted certificate. It will contain the old CA certificate, and the admin will make updates to in-memory trust-store based on updates generated by the Raft configuration via `nuocmd add/remove trusted-certificates`.

Reply:
> I've tested the steps with just 2 admins so I will double-check them with 3 to be sure.

CA certificate rotation with three APs is working fine as well.